src/sampling.jl

```diff
@@ -42,6 +42,47 @@ function direct_sample!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
 end
 direct_sample!(a::AbstractArray, x::AbstractArray) = direct_sample!(Random.GLOBAL_RNG, a, x)
 
+# check whether we can use T to store indices 1:n exactly
+for T in unique!([Int,UInt,Int64,UInt64,Int128,UInt128])
+    @eval isexact(n, ::Type{$T}) = true
+end
+isexact(n, ::Type{T}) where {T<:Integer} = n ≤ typemax(T)
+isexact(n, ::Type{BigInt}) = false # bigint is too expensive
+for T in (Float32, Float64)
+    @eval isexact(n, ::Type{$T}) = n ≤ maxintfloat($T)
+end
+isexact(n, ::Type{Complex{T}}) where {T} = isexact(n, T)
+isexact(n, ::Type{Rational{T}}) where {T} = isexact(n, T)
+isexact(n, T) = false
+
+# order results of a sampler that does not order automatically
+function sample_ordered!(sampler!::F, rng::AbstractRNG, a::AbstractArray, x::AbstractArray) where {F}
+    n, k = length(a), length(x)
+    if isexact(n, eltype(x))
+        sort!(sampler!(rng, Base.OneTo(n), x), by=real)
+        for i = 1:k
+            x[i] = a[Int(x[i])]
+        end
+    else
+        indices = Array{Int}(undef, k)
+        sort!(sampler!(rng, Base.OneTo(n), indices))
+        for i = 1:k
+            x[i] = a[indices[i]]
+        end
+    end
+    return x
+end
+
+# special case of a range (or any sorted array) can be done more efficiently
```

Comment:
Better not mention sorted arrays, as this method only applies to ranges. Would it make sense to call `issorted(x)` for all array types and use this fast path when it's true?

Reply:
`issorted` could throw if the array elements don't have a comparison.   But of course we could try `issorted` only for `<:Real` arrays, since it's just an optimization.

One worry is that if the array *isn't* sorted, it will actually be slower since it will pay the price of the `issorted` check in addition to what we are already doing.

Reply:
I have no idea, but in general I think `issorted` should be quite cheap since unsorted arrays are very unlikely to have there first elements sorted, so typically only a small proportion of them will be read I guess?

Reply:
I removed the comment here and added a note elsewhere in `sample_ordered!` that this might be a possible future optimization.